src/test/java/libai/common/MatrixTest.java

```diff
@@ -919,6 +928,62 @@ public double eval(double x) {
         }
     }
 
+	@Test(expected = IllegalArgumentException.class)
+	public void testDotProductThisNotRow() {
+		Matrix a = Matrix.random(2, 10);
+		Matrix b = Matrix.random(1, 20);
+		a.dotProduct(b);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testDotProductThisNotColumn() {
+		Matrix a = Matrix.random(10, 2);
+		Matrix b = Matrix.random(1, 20);
+		a.dotProduct(b);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testDotProductOtherNotRow() {
+		Matrix a = Matrix.random(1, 20);
+		Matrix b = Matrix.random(2, 10);
+		a.dotProduct(b);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testDotProductOtherNotColumn() {
+		Matrix a = Matrix.random(1, 20);
+		Matrix b = Matrix.random(10, 2);
+		a.dotProduct(b);
+	}
+
+	@Test
+	public void testHashCode() {
+		Matrix a = Matrix.random(10, 10);
+		Matrix b = Matrix.random(10, 10);
+		Matrix c = Matrix.random(10, 5);
+		Matrix d = Matrix.random( 5, 10);
+
+		HashMap<Matrix, Matrix> map = new HashMap<>(4);
+		map.put(a, a);
+		map.put(b, b);
+		map.put(c, c);
+		map.put(d, d);
+
+		assertTrue(a == map.get(a));
```

Comment:
i don't know until what extent this is actually testing anything. wouldn't it be better to create 2 matrices with the same values and ask if their hashcodes are the same?